File: src/prototype/check_adjoint.m
% check Boundary adjoint 

mX = rand(Q+1,N+2);
fX = rand(Q+2,N+1);

mY = rand(Q+1,2);
fY = rand(2,N+1);

b = boundary(mX,fX);

s1 = sum([sum(b.m.*mY) sum((b.f.*fY)')]);

[mA fA] = boundary_adjoint(mY(:,1),mY(:,2),fY(1,:),fY(2,:),N,Q);

s2 = sum([sum(mX.*mA) sum((fX.*fA)')]);

e1 = abs(s1 - s2)/s1

% check Divergence adjoint

mbarX = rand(Q+1,N+2);
fbarX = rand(Q+2,N+1);

dY = rand(Q+1,N+1);

dX = divergence(mbarX,fbarX);
[mbarY fbarY] = divergence_adjoint(dY);
s1 = sum(sum(dX.*dY));
s2 = sum([sum(mbarX.*mbarY) sum((fbarX.*fbarY)')]);

e2 = abs(s1 - s2)/s1

% check Interpolation adjoint
globals;
mbar = rand(Q+1,N+2); fbar = rand(Q+2,N+1);
m = rand(Q+1,N+1); f = rand(Q+1,N+1); 

[Imbar Ifbar] = interpolation(mbar,fbar);
[Iadjm Iadjf] = interpolation_adjoint(m,f); 

s1 = sum(sum(Imbar.*m)) + sum(sum(Ifbar.*f));
s2 = sum(sum(Iadjm.*mbar)) + sum(sum(Iadjf.*fbar));

e3 = abs(s1 - s2)/s1

File: src/prototype/Fortran/check_adjoint.f90
subroutine check_adjoint(N,Q)
    implicit none
    integer, intent(in) :: N,Q;

    !! Variables tests Boundary!!
    
    double precision, dimension(Q+1,N+2) :: mbar,boundAdjMres;
    double precision, dimension(Q+2,N+1) :: fbar,boundAdjFres;
    double precision, dimension(Q+1) :: mleft,mright,boundResLeft,boundResRight;
    double precision, dimension(N+1) :: fup,fdown,boundResUp,boundResDown;

    !! Variables pour l'interpolation !!
    
    double precision, dimension(Q+1,N+2) :: mbarInterpAdj;
    double precision, dimension(Q+2,N+1) :: fbarInterAdj;
    double precision, dimension(Q+1,N+1) :: m,f,mInterp,fInterp;

    !! Variables pour la divergence !! 

    double precision, dimension(Q+1,N+2) :: divAdjMres;
    double precision, dimension(Q+2,N+1) :: divAdjFres;
    double precision, dimension(Q+1,N+1) :: d,dRes;

    !! Variables pour tous le monde !!
    double precision :: s1,s2;


    !! Initialisation Boundary !!
    boundAdjMres = 0; boundAdjFres = 0;
    boundResLeft = 0; boundResRight = 0;
    boundResUp = 0; boundResDown = 0;
    call RANDOM_NUMBER(mbar);
    call RANDOM_NUMBER(fbar);
    call RANDOM_NUMBER(mleft);
    call RANDOM_NUMBER(mright);
    call RANDOM_NUMBER(fup);
    call RANDOM_NUMBER(fdown);

   
    print *, "Check Boundary adjoint";
    call boundary(mbar,fbar,boundResLeft,boundResRight,boundResUp,boundResDown,N,Q);
    call boundary_adjoint(boundAdjMres,boundAdjFres,mleft,mright,fup,fdown,N,Q);

    s1 = sum(boundResLeft*mleft) + sum(boundResRight*mright) + sum(boundResUp*fup) + sum(boundResDown*fdown);
    s2 = sum(boundAdjFres*fbar) + sum(boundAdjMres*mbar);

    print *, abs(s1-s2)/s1;

    !! Initialisation divergence !!
    divAdjMres = 0;
    divAdjFres = 0;
    d = 0;    
    call RANDOM_NUMBER(dRes);

    print *, "Check divergence adjoint";

    call divergence(mbar,fbar,d,N,Q);
    call divergence_adjoint(divAdjMres,divAdjFres,dRes,N,Q);

    s1 = sum(dRes*d);
    s2 = sum(divAdjMres*mbar) + sum(divAdjFres*fbar);

    print *, abs(s1-s2)/(1.0*s1);
    
 !! Initialisation Interpolation !!

    mbarInterpAdj = 0;
    fbarInterAdj = 0;
    mInterp = 0;
    fInterp = 0;
    call RANDOM_NUMBER(m);
    call RANDOM_NUMBER(f);

    print *, "Check interpolation adjoint";
  
    call interpolation(mbar,fbar,mInterp,fInterp,N,Q);
    call interpolation_adjoint(mbarInterpAdj,fbarInterAdj,m,f,N,Q);  

    s1 = sum(mInterp*m) + sum(fInterp*f);
    s2 = sum(mbarInterpAdj*mbar) + sum(fbarInterAdj*fbar);

    print *, abs(s1-s2)/(1.0*s1);

end subroutine